Source/Diagnostics/ReducedDiags/Costs.cpp

```diff
@@ -0,0 +1,199 @@
+
+    if (ParallelDescriptor::IOProcessor())
+    {
+        // final step is a special case, fill jagged array with NaN
+        if (step == (warpx.maxStep() - 1))
+        {
+            // open tmp file to copy data
+            std::string fileTmpName = m_path + m_rd_name + ".tmp." + m_extension;
+            std::ofstream ofs(fileTmpName, std::ofstream::out);
+
+            // write header row
+            // for each box on each level we saved 6 data fields: [cost, proc, lev, i_low, j_low, k_low])
+            ofs << "#";
+            ofs << "[1]step()";
+            ofs << m_sep;
+            ofs << "[2]time(s)";
+            ofs << m_sep;
+            for (int boxNumber=0; boxNumber<nBoxesMax; ++boxNumber)
+            {
+                ofs << "[" + std::to_string(3 + nDataFields*boxNumber) + "]";
+                ofs << "cost_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(4 + nDataFields*boxNumber) + "]";
+                ofs << "proc_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(5 + nDataFields*boxNumber) + "]";
+                ofs << "lev_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(6 + nDataFields*boxNumber) + "]";
+                ofs << "i_low_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(7 + nDataFields*boxNumber) + "]";
+                ofs << "j_low_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(8 + nDataFields*boxNumber) + "]";
+                ofs << "k_low_box_"+std::to_string(boxNumber)+"()";
+                ofs << m_sep;
+            }
+            ofs << std::endl;
+
+            // open the data-containing file
+            std::string fileDataName = m_path + m_rd_name + "." + m_extension;
+            std::ifstream ifs(fileDataName, std::ifstream::in);
+
+            // Fill in the tmp costs file with data, padded with NaNs
+            for (std::string lineIn; std::getline(ifs, lineIn);)
+            {
+                // count the elements in the input line
+                int cnt = 0;
+                std::stringstream ss(lineIn);
+                std::string token;
+
+                while (std::getline(ss, token, m_sep[0]))
+                {
+                    cnt += 1;
+                    if (ss.peek() == m_sep[0]) ss.ignore();
+                }
+
+                // 2 columns for step, time; then nBoxes*nDatafields columns for data;
+                // then fill the remaining columns (i.e., up to 2 + nBoxesMax*nDataFields)
+                // with NaN, so the array is not jagged
+                ofs << lineIn;
+                for (int i=0; i<(nBoxesMax*nDataFields - (cnt - 2)); ++i)
+                {
+                    ofs << m_sep << "NaN";
```

Comment:
Could there be a case one of your costs is zero? If not, maybe filling zeros is better than "Nan"?